### ISSUE TITLE
Resizing results in low quality/blurry image

### ISSUE THREAD

**Issue by crypto-design-club** [lovell/sharp#4185]
## Possible bug

Resizing images produces low-quality/blurry images. I tried using different downsizing kernels but the results were consistently bad.

FYI I'm getting to this code looking through the Strapi upload plugin code and https://github.com/nicolashmln/strapi-plugin-responsive-image, but it also happens when I try to resize an image standalone with the code sample provided below.

### Is this a possible bug in a feature of sharp, unrelated to installation?

- [x] Running `npm install sharp` completes without error.
- [x] Running `node -e "require('sharp')"` completes without error.

If you cannot confirm both of these, please open an [installation issue](https://github.com/lovell/sharp/issues/new?labels=installation&template=installation.md) instead.

### Are you using the latest version of sharp?

- [x] I am using the latest version of `sharp` as reported by `npm view sharp dist-tags.latest`.

If you cannot confirm this, please upgrade to the latest version and try again before opening an issue.

If you are using another package which depends on a version of `sharp` that is not the latest, please open an issue against that package instead.

### What is the output of running `npx envinfo --binaries --system --npmPackages=sharp --npmGlobalPackages=sharp`?

  System:
    OS: macOS 14.5
    CPU: (8) arm64 Apple M3
    Memory: 934.55 MB / 24.00 GB
    Shell: 5.9 - /bin/zsh
  Binaries:
    Node: 22.4.1 - /opt/homebrew/bin/node
    Yarn: 1.22.22 - /opt/homebrew/bin/yarn
    npm: 10.8.1 - /opt/homebrew/bin/npm
    bun: 1.1.18 - ~/.bun/bin/bun
  npmPackages:
    sharp: ^0.33.4 => 0.33.4

### Does this problem relate to file caching?

The default behaviour of libvips is to cache input files, which can lead to `EBUSY` or `EPERM` errors on Windows.
Use [`sharp.cache(false)`](https://sharp.pixelplumbing.com/api-utility#cache) to switch this feature off.

- [x] Adding `sharp.cache(false)` does not fix this problem.

### Does this problem relate to images appearing to have been rotated by 90 degrees?

Images that contain EXIF Orientation metadata are not auto-oriented. By default, EXIF metadata is removed.

- To auto-orient pixel values use the parameter-less [`rotate()`](https://sharp.pixelplumbing.com/api-operation#rotate) operation.
- To retain EXIF Orientation use [`keepExif()`](https://sharp.pixelplumbing.com/api-output#keepexif).

- [x] Using `rotate()` or `keepExif()` does not fix this problem.

### What are the steps to reproduce?

1. Copy the code sample below into a .js file
2. Create a folder images and place the test image there, call it test.jpeg
3. Created a folder processed_images
4. Run with node yourfile.js
5. Compare the output image with the input, the quality is way lower now

### What is the expected behaviour?

The output image quality should be about equal than the input, since it only does resizing.

### Please provide a minimal, standalone code sample, without other dependencies, that demonstrates this problem

```
const sharp = require('sharp')
const fs = require('fs')
const path = require('path')

const writeStreamToFile = (stream, path) =>
  new Promise((resolve, reject) => {
    const writeStream = fs.createWriteStream(path);
    // Reject promise if there is an error with the provided stream
    stream.on("error", reject);
    stream.pipe(writeStream);
    writeStream.on("close", resolve);
    writeStream.on("error", reject);
  });

const resize = async () => {
  const fileStream = fs.createReadStream(path.join(__dirname, 'images', 'test.jpeg'))
  const filePath = 'processed_images/test.jpeg'
  const sharpInstance = sharp()
  sharpInstance.resize({
    width: 1000,
  });
  await writeStreamToFile(fileStream.pipe(sharpInstance), filePath);
}

resize()
```

### Please provide sample image(s) that help explain this problem

[here](https://drive.google.com/file/d/1uzdq38cdBnZJaY-6cGvYsZJkczurceV-/view?usp=sharing)

**Comment by lovell** [comment]
If you want JPEG output, which is always lossy, the `quality` parameter provides control over this.

https://sharp.pixelplumbing.com/api-output#jpeg

```js
const sharpInstance = sharp()
  .resize({
    width: 1000,
  })
  .jpeg({
    quality: 100,
  })
```

The sample image provided is a screenshot and would probably be better served using a lossless image format such as PNG.

**Comment by crypto-design-club** [comment]
@lovell Thanks for coming back so quickly!

I had a try with .jpeg({ quality: 100 }) and it didn't go much better.

Then I tried to follow your advice and converted the image into a PNG and ran my script again, but still no luck:

```
const sharpInstance = sharp()
  .resize({
    width: 1000,
  })
  .png({
    quality: 100, // I tried without this setting too, just resize({ width: 1000 })
  })
```

Please find the [PNG image here](https://drive.google.com/file/d/1iWefRMGj3PnrnMj9ViCxahyE98FilPFG/view?usp=sharing), the original seems quite sharp to me, but the output unfortunately not when viewed in its proper size (1000px wide) in Mac Preview or in my Chrome.

I don't know how this library works internally but I guess it is because there is quite a lot of small text?

**Comment by lovell** [comment]
If you want small, light text on a dark background to "pop" more, perhaps try adding [sharpen](https://sharp.pixelplumbing.com/api-operation#sharpen) to your pipeline.

**Comment by crypto-design-club** [comment]
@lovell I tried adding sharpen to it, but didn't really help. It's not so much about making the text pop more vs. the whole image being blurry after resizing...

I tried adding `sharpen` without parameters and with some variation (give by GPT):

```
  const sharpInstance = sharp()
  sharpInstance.resize({
    width: 1000,
    kernel: 'lanczos3',
    withoutEnlargement: true,
  })
  .sharpen({
    sigma: 1,
    m1: 1,
    m2: 2,
    x1: 0,
    y2: 1.5,
    y3: 2
  })
  .png({
    quality: 100,
    compressionLevel: 9,
    adaptiveFiltering: true,
  })
```

Unfortunately, nothing I try seems to make a big difference... If it's of any help I'm using CleanShot X now taking screenshots and they appear really sharp in a much greater resolution. Not sure what to try next to resize this properly, seems like I might have to use a different resizing tool and only do the image conversion with sharp then. Any guidance here would be appreciated as I assumed resizing to be a standard operation and to work out of the box for many different screenshots I intend to upload on my CMS.

**Comment by crypto-design-club** [comment]
Closing this one, it's not a sharp issue, I couldn't get this resized with other tools either. Downsizing screenshots containing fonts seems to be a general issue, so closing it here.

**Comment by mertalev** [comment]
[This](https://stackoverflow.com/a/13243833) post might help with finding the right resizing kernel for the job. Sharp doesn't support as many as ImageMagick, so ImageMagick might be better for this.

There is also [Magic Kernel Sharp](https://www.johncostella.com/magic/), which looks very promising.

**Comment by j-p-c** [comment]
I can't see what your output was using this library, but since Magic Kernel Sharp was mentioned above, this is what I get when I resize your image using it.
![test-1000](https://github.com/user-attachments/assets/b51d8abe-882a-4bb3-8e7e-680e84287fbb)

**Comment by boris-szl** [comment]
What is the status on this? @crypto-design-club have you been able to solve this? We are facing the same issue. We have a photo with a timestamp that is added after the images is captured. Resizing the image makes the timestamp unreadable .We need to resize the images in a report generation context.

**Comment by j-p-c** [comment]
> What is the status on this? [@crypto-design-club](https://github.com/crypto-design-club) have you been able to solve this? We are facing the same issue. We have a photo with a timestamp that is added after the images is captured. Resizing the image makes the timestamp unreadable .We need to resize the images in a report generation context.

I am happy to help you implement [Magic Kernel Sharp](https://johncostella.com/magic/) if it hasn't already been implemented.

**Comment by lovell** [comment]
Resizing images using _Magic Kernel Sharp_ 2013 and 2021 has been available since v0.34.3

- https://sharp.pixelplumbing.com/api-resize/
- https://sharp.pixelplumbing.com/changelog/v0.34.3/

**Comment by j-p-c** [comment]
> Resizing images using _Magic Kernel Sharp_ 2013 and 2021 has been available since v0.34.3
> 
> * https://sharp.pixelplumbing.com/api-resize/
> * https://sharp.pixelplumbing.com/changelog/v0.34.3/

Ah, my bad—sorry.

@boris-szl : Are you using MKS?

**Comment by boris-szl** [comment]
Appreciate the quick reply! @lovell @j-p-c Thank you!
                                                                                                                                                                                                            
We tested all three kernels (lanczos3, mks2013, mks2021) at multiple dimensions. At 100% zoom on the raw output, mks2013 produces the sharpest result — consistent with its more aggressive sharpening. But in our actual PDF context (images rendered at ~350px CSS height on A4 pages), the differences between all three kernels are visually indistinguishable.                                               

For the timestamp readability issue specifically — the kernel choice alone won't solve it. The real lever is output dimensions. If the timestamp text is small relative to the image, no downscaling kernel can preserve it — it's an information density problem. You might get better results by increasing output dimensions or rendering the timestamp at a larger font size before capture.             

Our pipeline:                                                                                                                                                                                           

```
  const optimized = await sharp(imageBuffer)
      .resize(800, 600, {
          withoutEnlargement: true,
          fit: sharp.fit.inside,                                                                                                                                                                            
          kernel: sharp.kernel.mks2013,
      })                                                                                                                                                                                                    
      .jpeg({                                                                                                                                                                                             
          quality: 100,
          mozjpeg: true,
      })
      .toBuffer();
```

We went with mks2013 as it's the sharpest of the three and the proven standard (Facebook/Instagram). mozjpeg: true and quality: 100 because downstream compression compounds any artifacts introduced here.

I can drop in a raw source image if you'd like to test against it or have further recommendations on the pipeline.                                                                                      

Best regards!
